pylace/src/update_handler.rs

```diff
@@ -0,0 +1,139 @@
+use std::io::Write;
+/// Update Handler and associated tooling for `CoreEngine.update` in Python.
+use std::sync::{Arc, Mutex};
+
+use lace::cc::state::State;
+use lace::update_handler::UpdateHandler;
+use lace::EngineUpdateConfig;
+use pyo3::{pyclass, IntoPy, Py, PyAny};
+
+/// Python version of `EngineUpdateConfig`.
+#[derive(Clone, Debug)]
+#[pyclass(frozen, get_all)]
+pub struct PyEngineUpdateConfig {
+    /// Maximum number of iterations to run.
+    pub n_iters: usize,
+    /// Number of iterations after which each state should be saved
+    pub checkpoint: Option<usize>,
+    /// Number of states
+    pub n_states: usize,
+}
+
+/// An `UpdateHandler` which wraps a Python Object.
+#[derive(Debug, Clone)]
+pub struct PyUpdateHandler {
+    handler: Arc<Mutex<Py<PyAny>>>,
+}
+
+impl PyUpdateHandler {
+    /// Create a new `PyUpdateHandler` from a Python Object
+    pub fn new(handler: Py<PyAny>) -> Self {
+        Self {
+            handler: Arc::new(Mutex::new(handler)),
+        }
+    }
+}
+
+macro_rules! pydict {
+    ($py: expr, $($key:tt : $val:expr),* $(,)?) => {{
+
+        let map = pyo3::types::PyDict::new($py);
+        $(
+            let _ = map.set_item($key, $val.into_py($py))
+                .expect("Should be able to set item in PyDict");
+        )*
+        map
+    }};
+}
+
+macro_rules! call_pyhandler_noret {
+    ($self: ident, $func_name: tt, $($key: tt : $val: expr),* $(,)?) => {{
+        let handler = $self
+            .handler
+            .lock()
+            .expect("Should be able to get a lock for the PyUpdateHandler");
```

Comment:
I'm not 100% sure what happens if you panic inside a Rayon thread. Will it do what we want (i.e., abort all execution)?

Reply:
Yes, it should panic and bring down the whole process via the standard panic handler.

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=a33a4109f7897af06cac6e083f8cf76a